src/main/kotlin/org/rust/cargo/project/workspace/impl/CargoProjectWorkspaceImpl.kt

```diff
@@ -138,54 +154,102 @@ class CargoProjectWorkspaceImpl(private val module: Module) : CargoProjectWorksp
                     }
                 })
 
-                Result.Ok(description)
+                UpdateResult.Ok(description)
             } catch (e: ExecutionException) {
-                Result.Err(e)
+                UpdateResult.Err(e)
             }
         }
 
         override fun onSuccess() {
-            val result = requireNotNull(result)
+            val r = requireNotNull(result)
 
-            when (result) {
-                is Result.Err -> {
-                    showBalloon("Project '${module.project.name}' update failed: ${result.error.message}", MessageType.ERROR)
-                    LOG.info("Project '${module.project.name}' update failed", result.error)
+            lock (lock) {
+                cached = when (r) {
+                    is UpdateResult.Ok -> r.projectDescription
+                    else               -> null
                 }
-                is Result.Ok  ->
+
+                notifyCargoProjectUpdate(r)
+            }
+
+            when (r) {
+                is UpdateResult.Ok  -> LOG.info("Project '${module.project.name}' successfully updated")
+                is UpdateResult.Err -> LOG.info("Project '${module.project.name}' update failed", r.error)
+            }
+        }
+    }
+
+    /**
+     * Project-update task's notifier
+     */
+    inner class Notifier : CargoProjectWorkspaceListener {
```

Comment:
Original motivation for the previous design is in this commit message: https://github.com/intellij-rust/intellij-rust/commit/e7b177b129095ed302d08cad192bd96920b4e12d